theano/tensor/nnet/abstract_conv.py

```diff
@@ -177,6 +179,383 @@ def conv2d(input,
+                                                                    1),
+                                               output_grad_shape=up_input.shape,
+                                               input_shape=(bch * ch, 1,
+                                                            row * ratio - sub,
+                                                            col),
+                                               filters_shape=(1, 1, kern_size,
+                                                              1),
+                                               border_mode='valid',
+                                               subsample=(1, 1),
+                                               filter_flip=True)
+        # upsampling cols
+        upsampled_mat = conv2d_grad_wrt_inputs(output_grad=upsampled_row,
+                                               filters=kern.reshape(1, 1, 1,
+                                                                    kern_size),
+                                               output_grad_shape=upsampled_row.shape,
+                                               input_shape=(bch * ch, 1,
+                                                            row * ratio - sub,
+                                                            col * ratio - sub),
+                                               filters_shape=(1, 1, 1,
+                                                              kern_size),
+                                               border_mode='valid',
+                                               subsample=(1, 1),
+                                               filter_flip=True)
+    else:
+        kern = bilinear_kernel_2D(ratio=ratio, normalize=True)
+        kern_row, kern_col = kern.shape
+        upsampled_mat = conv2d_grad_wrt_inputs(output_grad=up_input,
+                                               filters=kern.reshape(1, 1,
+                                                                    kern_row,
+                                                                    kern_col),
+                                               output_grad_shape=up_input.shape,
+                                               input_shape=(bch * ch, 1,
+                                                            row * ratio - sub,
+                                                            col * ratio - sub),
+                                               filters_shape=(1, 1, kern_row,
+                                                              kern_col),
+                                               border_mode='valid',
+                                               subsample=(1, 1),
+                                               filter_flip=True)
+
+    if sub:
+        # ratio is even
+        up_bch, up_ch, up_row, up_col = upsampled_mat.shape
+        if repeat_first_row:
+            upsampled_mat = T.concatenate((upsampled_mat[:, :, 0, :].reshape((
```

Comment:
Please use the much cleaner `T.concatenate((upsampled_mat[:, :, 0:1, :], upsampled_mat), axis=2)`